onedal/utils/_array_api.py

```diff
@@ -0,0 +1,91 @@
+# ==============================================================================
+# Copyright 2024 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+"""Tools to support array_api."""
+
+from collections.abc import Iterable
+
+try:
+    from dpctl.tensor import usm_ndarray
+
+    dpctl_available = True
+except ImportError:
+    dpctl_available = False
+
+try:
+    import dpnp
+
+    dpnp_available = True
+except ImportError:
+    dpnp_available = False
+
+
+if dpnp_available:
+    import dpnp
+
+    def _convert_to_dpnp(array):
+        """Converted input object to dpnp.ndarray format."""
+        # Will be removed and `onedal.utils._array_api._asarray` will be
+        # used instead after DPNP Array API enabling.
+        if isinstance(array, usm_ndarray):
+            return dpnp.array(array, copy=False)
+        elif isinstance(array, Iterable):
+            for i in range(len(array)):
+                array[i] = _convert_to_dpnp(array[i])
+        return array
+
+
+def _asarray(data, xp, *args, **kwargs):
+    """Converted input object to array format of xp namespace provided."""
+    if hasattr(data, "__array_namespace__"):
+        return xp.asarray(data, xp, *args, **kwargs)
+    elif isinstance(data, Iterable):
+        if isinstance(data, tuple):
+            result_data = []
+            for i in range(len(data)):
+                result_data.append(_asarray(data[i], xp, *args, **kwargs))
+            data = tuple(result_data)
+        else:
+            for i in range(len(data)):
+                data[i] = _asarray(data[i], xp, *args, **kwargs)
```

Comment:
Not is the best impl. Just need to check tests first. Will refactor.

Reply:
At the moment, the implementation passes all tests. If there is a request for refactoring, then it can be done separately so as not to delay the merge of the main feature and overload this PR.

Reply:
please make a ticket so we can cover this use case: https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestNeighbors.html#sklearn.neighbors.NearestNeighbors.radius_neighbors but I will agree with a later refactor.

Reply:
Let me cover this with the ticket and check it.

Reply:
Enable tests from [scikit-learn-intelex](https://github.com/intel/scikit-learn-intelex/tree/main)/[sklearnex](https://github.com/intel/scikit-learn-intelex/tree/main/sklearnex)/[tests](https://github.com/intel/scikit-learn-intelex/tree/main/sklearnex/tests)
/test_patching.py for array_api inputs

Reply:
Mentioned on 8366